File: packages/android-client/FishjamClient/src/main/java/com/fishjamcloud/client/Command.kt
package com.fishjamcloud.client

import kotlinx.coroutines.CompletableJob
import kotlinx.coroutines.Job

internal enum class CommandName {
  CONNECT,
  JOIN,
  ADD_TRACK,
  REMOVE_TRACK,
  RENEGOTIATE,
  LEAVE
}

internal enum class ClientState {
  CREATED,
  CONNECTED,
  JOINED
}

internal data class Command(
  val commandName: CommandName,
  val clientStateAfterCommand: ClientState?,
  val execute: () -> Unit = {}
) {
  constructor(commandName: CommandName, command: () -> Unit = {}) : this(commandName, null, command)

  val job: CompletableJob = Job()
}

File: packages/ios-client/Sources/FishjamClient/Command.swift
import Promises

public enum CommandName {
    case CONNECT, JOIN, ADD_TRACK, REMOVE_TRACK, RENEGOTIATE, LEAVE
}

public enum ClientState {
    case CREATED, CONNECTED, JOINED
}

public class Command {
    let commandName: CommandName
    let clientStateAfterCommand: ClientState?
    let promise: Promise<Void>
    let block: () -> Void

    public init(commandName: CommandName, clientStateAfterCommand: ClientState?, block: @escaping () -> Void) {
        self.commandName = commandName
        self.clientStateAfterCommand = clientStateAfterCommand
        self.block = block
        self.promise = Promise<Void>.pending()
    }

    func execute() {
        DispatchQueue.fishjam.async {
            self.block()
            self.promise.fulfill(())
        }
    }
}
